components/codec/buffer.rs

```diff
@@ -0,0 +1,435 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// A trait to provide sequential read over a memory buffer.
+///
+/// The memory buffer can be `&[u8]` or `std::io::Cursor<AsRef<[u8]>>`.
+pub trait BufferReader {
+    /// Returns a slice starting at current position.
+    ///
+    /// The returned slice can be empty.
+    fn bytes(&self) -> &[u8];
+
+    /// Advances the position of internal cursor.
+    ///
+    /// # Panics
+    ///
+    /// This function may panic in some implementors when remaining space
+    /// is not large enough to advance.
+    fn advance(&mut self, count: usize);
+}
+
+impl<T: AsRef<[u8]>> BufferReader for ::std::io::Cursor<T> {
+    fn bytes(&self) -> &[u8] {
+        let pos = self.position() as usize;
+        let slice = self.get_ref().as_ref();
+        if pos >= slice.len() {
+            return &[];
+        }
+        &slice[pos..]
+    }
+
+    fn advance(&mut self, count: usize) {
+        let mut pos = self.position();
+        pos += count as u64;
+        self.set_position(pos);
+    }
+}
+
+impl<'a> BufferReader for &'a [u8] {
+    fn bytes(&self) -> &[u8] {
+        self
+    }
+
+    fn advance(&mut self, count: usize) {
+        *self = &self[count..]
+    }
+}
+
+impl<'a, T: BufferReader + ?Sized> BufferReader for &'a mut T {
+    fn bytes(&self) -> &[u8] {
+        (**self).bytes()
+    }
+
+    fn advance(&mut self, count: usize) {
+        (**self).advance(count)
+    }
+}
+
+impl<T: BufferReader + ?Sized> BufferReader for Box<T> {
+    fn bytes(&self) -> &[u8] {
+        (**self).bytes()
+    }
+
+    fn advance(&mut self, count: usize) {
+        (**self).advance(count)
+    }
+}
+
+/// A trait to provide sequential write over a fixed size
+/// or dynamic size memory buffer.
+///
+/// The memory buffer can be `std::io::Cursor<AsRef<[u8]>>` or `&mut [u8]`,
+/// which is fixed sized, or `Vec<u8>`, which is dynamically sized.
+pub trait BufferWriter {
+    /// Returns a mutable slice starting at current position.
+    ///
+    /// The caller may hint the underlying buffer to grow according to `size`
+    /// if the underlying buffer is dynamically sized (i.e. is capable to grow).
+    ///
+    /// The size of the returned slice may be less than `size` given. For example,
+    /// when underlying buffer is fixed sized and there is no enough space any more.
+    ///
+    /// The returned mutable slice is for writing only and should be never used for
+    /// reading since it might contain uninitialized memory when underlying buffer
+    /// is dynamically sized. For this reason, this function is marked `unsafe`.
+    unsafe fn bytes_mut(&mut self, size: usize) -> &mut [u8];
```

Comment:
Why not return a struct that mutate the position when dropped? So `unsafe` is not necessary anymore.

Reply:
The un-safety of `bytes_mut` is that returned slice may contain uninitialized memory data, which should only be used to overwrite but not read. Seems to be unavoidable.

Reply:
> ...should only be used to overwrite but not read...

That's why I suggest to return a struct. In that way, the struct can provide a safe interface for users.

Reply:
But you cannot restrict the user not reading from the returned struct. It is still unsafe.

Reply:
Why preserve the slice interface? There should be only append operations for such buffer. In such case, only `push` and `extend_from_slice` is necessary.